packages/kbn-unified-data-table/src/components/data_table.tsx

```diff
@@ -603,74 +604,42 @@ export const UnifiedDataTable = ({
     typeof rowsPerPageState === 'number' && rowsPerPageState > 0
       ? rowsPerPageState
       : DEFAULT_ROWS_PER_PAGE;
-  const [pagination, setPagination] = useState({
-    pageIndex: 0,
-    pageSize: currentPageSize,
-  });
+  const [currentPageIndex, setCurrentPageIndex] = useState(0);
   const rowCount = useMemo(() => (displayedRows ? displayedRows.length : 0), [displayedRows]);
   const pageCount = useMemo(
-    () => Math.ceil(rowCount / pagination.pageSize),
-    [rowCount, pagination]
+    () => Math.ceil(rowCount / currentPageSize),
+    [currentPageSize, rowCount]
   );
 
-  const paginationObj = useMemo(() => {
+  const pagination = useMemo<EuiDataGridPaginationProps | undefined>(() => {
     const onChangeItemsPerPage = (pageSize: number) => {
       onUpdateRowsPerPage?.(pageSize);
     };
 
     const onChangePage = (pageIndex: number) => {
-      setPagination((paginationData) => ({ ...paginationData, pageIndex }));
+      setCurrentPageIndex(pageIndex);
+      onUpdatePageIndex?.(pageIndex);
```

Comment:
```suggestion
      const calculatedPageIndex = pageIndex > pageCount - 1 ? 0 : pageIndex
      setCurrentPageIndex(calculatedPageIndex);
      onUpdatePageIndex?.(calculatedPageIndex);
```
I think this implementation could be updated to fix the bug you found by calculating the page index within this callback...

Reply:
@davismcphee , i do not think this would work, because `onChangePage` is not called when we need new `pageIndex` to be calculate. Let's have a short call today.

Reply:
Also, I think `onUpdatePageIndex` should now be solely based on `currentPageIndex`. Any changes to that should be automatically trigger `onUpdatePageIndex` instead of calling `onUpdatePageIndex` at multiple places.